src/bin/sccache-dist/cmdline/parse.rs

```diff
@@ -0,0 +1,369 @@
+use std::{env, ffi::OsString, fmt, net::SocketAddr, path::PathBuf, str::FromStr};
+
+use anyhow::bail;
+use clap::{Arg, ArgEnum, ArgGroup, Command as ClapCommand, PossibleValue};
+use sccache::{config, dist::ServerId};
+use syslog::Facility;
+
+use crate::cmdline::{AuthSubcommand, Command};
+
+#[derive(Debug)]
+struct TokenBits(usize);
+
+impl TokenBits {
+    fn as_bytes(&self) -> usize {
+        self.0 / 8
+    }
+
+    fn default_str() -> &'static str {
+        // Size based on https://briansmith.org/rustdoc/ring/hmac/fn.recommended_key_len.html
+        "256"
+    }
+}
+
+impl Default for TokenBits {
+    fn default() -> Self {
+        Self::default_str().parse().expect("Default isn't valid")
+    }
+}
+
+impl FromStr for TokenBits {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bits: usize = s.parse()?;
+        if bits % 8 != 0 || bits < 64 || bits > 4_096 {
+            bail!("Number of bits must be divisible by 8, greater than 64 and less than 4096")
+        }
+
+        Ok(Self(bits))
+    }
+}
+
+impl fmt::Display for TokenBits {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(ArgEnum, Clone, Copy)]
+enum LogLevel {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl LogLevel {
+    fn possible_values() -> impl Iterator<Item = PossibleValue<'static>> {
+        Self::value_variants()
+            .iter()
+            .filter_map(ArgEnum::to_possible_value)
+    }
+}
+
+impl FromStr for LogLevel {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let variant = match s {
+            "error" => Self::Error,
+            "warn" => Self::Warn,
+            "info" => Self::Info,
+            "debug" => Self::Debug,
+            "trace" => Self::Trace,
+            _ => bail!("Unknown log level: {:?}", s),
+        };
+
+        Ok(variant)
+    }
+}
+
+impl From<LogLevel> for log::LevelFilter {
+    fn from(log_level: LogLevel) -> Self {
+        match log_level {
+            LogLevel::Error => Self::Error,
+            LogLevel::Warn => Self::Warn,
+            LogLevel::Info => Self::Info,
+            LogLevel::Debug => Self::Debug,
+            LogLevel::Trace => Self::Trace,
+        }
+    }
+}
+
+/// Parse the commandline into a `Command` to execute.
+pub fn parse() -> Command {
+    match try_parse_from(env::args()) {
+        Ok(cmd) => cmd,
+        Err(e) => match e.downcast::<clap::error::Error>() {
+            Ok(clap_err) => clap_err.exit(),
+            Err(some_other_err) => {
+                println!("sccache-dist: {some_other_err}");
+                for source_err in some_other_err.chain().skip(1) {
+                    println!("sccache-dist: caused by: {source_err}");
+                }
+                std::process::exit(1);
```

Comment:
Let's move the interface between our code and the CLI (error output, `sys::process:exit(1)`) up one level to sit in `sccache-dist/main.rs` again.

Reply:
Sounds good. Having a function that may exit without having an obvious name doesn't seem like the best idea. Moving it back into `sccache-dist/main.rs` reduces that separation :+1:

Reply:
Moved all the functionality that would `exit` back into the binaries respective `main`s in 87ed6e8f455e0e65e6e64ca8cd844ae72c9c7419